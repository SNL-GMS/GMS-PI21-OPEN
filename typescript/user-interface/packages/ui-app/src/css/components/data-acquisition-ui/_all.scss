@import 'soh-vars';
@import 'soh-cell';
@import 'ag-theme-overrides.scss';
@import 'data-acquisition';
@import 'environment-history';
@import 'historical-trends-history';
@import 'soh-chart';
@import 'station-statistics';
@import 'soh-lag-missing';
@import 'soh-environment';
@import 'soh-overview';
@import 'soh-quiet-indicator';
@import 'capability-counts';
@import 'soh-map';
