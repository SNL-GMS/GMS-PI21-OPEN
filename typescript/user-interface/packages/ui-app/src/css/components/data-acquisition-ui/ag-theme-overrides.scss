.ag-theme-dark--soh {
  border-radius: $cell-cap-radius;

  .ag-header {
    background: transparent;
  }

  .ag-header-cell {
    .ag-react-container {
      width: 100%;
    }
  }
  .ag-header-viewport {
    background-color: var(--gms-prominent-background, $gms-prominent-background);
  }

  .ag-row {
    overflow: hidden;
    border-radius: $cell-cap-radius;
  }

  .ag-ltr .ag-cell:not(.ag-cell-focus).ag-cell-last-left-pinned,
  .ag-ltr .ag-row.ag-cell-last-left-pinned,
  .ag-horizontal-left-spacer {
    border-right: none;
  }

  &:not(.table--cell-selection-only) {
    .ag-row-hover .ag-cell {
      @include brighten;
    }
  }
  &.table--cell-selection-only {
    .ag-cell:hover {
      @include brighten;
    }
  }

  // TODO: Make this dynamic so it doesn't expect a fixed
  // number of columns
  .ag-row-selected {
    .ag-cell .table-cell {
      box-sizing: border-box;
      @include selectedBorder('top', 'bottom');
    }
    .ag-cell[aria-colindex='1'] .table-cell {
      @include selectedBorder('left');
    }
    .ag-cell .table-cell[data-col-position='last'] {
      @include selectedBorder('right');
    }
  }

  .ag-body-viewport,
  .ag-root-wrapper {
    background: transparent;
  }
}
