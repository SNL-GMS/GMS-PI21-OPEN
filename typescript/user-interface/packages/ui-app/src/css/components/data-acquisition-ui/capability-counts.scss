.stacked-counts {
  @include flexColumn;
  @include centerChildren;
  &__entry {
    padding: 0;
    margin: 0;
    line-height: 1em;
    padding-right: 0.25rem;
    margin-left: -0.25rem;
    font-weight: 400;
    opacity: 0.8;
    font-size: 0.875em;
    & + & {
      margin-top: 0.125rem;
    }
  }
}

// Color the marginal capability count yellow if it is inside of a red cell
[data-cell-status='bad'] .stacked-counts__entry[data-capability-status='marginal'] {
  color: $marginal-color;
}
