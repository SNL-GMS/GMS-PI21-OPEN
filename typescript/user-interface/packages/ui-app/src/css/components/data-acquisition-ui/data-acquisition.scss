.form-field-small {
  width: 60%;
}

.form-field-large {
  margin-left: 5%;
  width: 80%;
}

.footer-button-row {
  position: absolute;
  top: 90%;
  left: 1.5%;
}

.footer-button {
  position: relative;
  display: inline;
}

.footer-button-middle {
  position: relative;
  display: inline;
  left: 200%;
}

.footer-button-right {
  position: relative;
  display: inline;
  left: 225%;
}

.station-configuration-form__row {
  display: flex;
  flex-direction: row;
  margin-bottom: 1%;
}

.station-configuration-form {
  display: flex;
  position: absolute;
  top: 0%;
  height: 20%;
  right: 0%;
  left: 0.5%;
  flex-direction: column;
  width: 100%;
  overflow-y: scroll;
}

.station-configuration-inner-menu {
  position: absolute;
  top: 45%;
  height: 20%;
  right: 0%;
  left: 1.5%;
}

.station-configuration-bottom-menu {
  position: absolute;
  top: 68%;
  height: 20%;
  right: 0%;
  left: 1.5%;
}

.station-configuration-form__labels-top-left {
  margin-left: 1% !important;
  margin-top: 0.5% !important;
  margin-right: 1% !important;
  font-size: small;
}

.station-configuration-form__labels-top-right {
  margin-left: 1% !important;
  margin-top: 1% !important;
  font-size: small;
}

.station-configuration-form__labels-middle {
  margin-left: 1% !important;
  margin-right: 15px !important;
  font-size: small;
}

.station-configuration-form__field-top {
  margin-left: 1.5% !important;
  margin-top: 15px !important;
  font-size: small;
}

.station-configuration-form__field-middle {
  margin-left: 12px !important;
  margin-top: 5px !important;
  font-size: small;
  width: 200%;
}

.station-configuration-form__field-top {
  margin-left: 5px !important;
  margin-top: 5px !important;
  font-size: small;
}

.station-configuration-form__field-bottom {
  margin-left: 3px !important;
  margin-top: 5px !important;
  font-size: small;
}

.station-configuration-form__labels-bottom-right {
  margin-left: 1% !important;
  font-size: small;
}

.station-configuration-form__labels-bottom-left {
  margin-left: 1% !important;
  font-size: small;
}

.station-configuration-form__input-middle {
  padding-left: 10% !important;
  width: 200%;
}

.configure-station-groups-select-network {
  position: relative;
  top: 0%;
  height: 100%;
  width: 40%;
  right: 0%;
  left: 0%;
}

.configure-station-groups-select-network-table {
  position: relative;
  top: 0%;
  height: 100%;
  max-height: calc(100% - 49px);
  width: 100%;
  right: 0%;
  left: 0%;
}

.configure-station-groups-edit-stations {
  position: absolute;
  top: 0%;
  height: 98%;
  right: 0%;
  left: 41%;
  visibility: hidden;
}

.configure-station-groups-no-network-selected {
  position: absolute;
  top: 0%;
  height: 98%;
  right: 0%;
  left: 41%;
  visibility: visible;
}

.configure-station-groups-associated-stations {
  position: absolute;
  top: 0%;
  height: 100%;
  width: 45%;
  right: 0%;
  left: 0%;
}

.configure-station-groups-arrows {
  position: absolute;
  top: 40%;
  height: 100%;
  width: 4%;
  right: 0%;
  left: 48%;
}

.configure-station-groups-available-stations {
  position: absolute;
  top: 0%;
  height: 100%;
  width: 45%;
  right: 0%;
  left: 55%;
}

.data-acquisition-popover {
  font-size: small !important;
  max-height: 300px !important;
  overflow-y: auto !important;
  margin: 0.5rem;
  display: inline-table;
}

.data-acquisition-import-css-popover-with-table {
  font-size: small !important;
  max-height: 500px !important;
  overflow-y: scroll !important;
  margin: 0.5rem;
}

.data-acquisition-inputs input {
  border: 1px solid var(--gms-main, $gms-main);
}

.data-acquisition-inputs input:focus {
  outline: none !important;
  border: 1px solid $data-acq-strong-warning;
}

.data-acquisition-cell .ag-cell {
  display: flex;
  justify-content: center;
  align-items: center !important;
}

.soh-update-time-controls {
  flex: 0 0 auto;
  width: 100%;
  display: flex;
  justify-content: space-around;
  align-items: center;
  color: var(--gms-main, $gms-main) !important;
}

.soh-update-time-controls__label {
  margin-bottom: 0rem !important;
  margin-left: 1rem !important;
  color: var(--gms-main, $gms-main);
}

.soh-update-time {
  font-family: $gms-mono;
  color: var(--gms-main, $gms-main);
}

.text-opacity-80 {
  opacity: 0.8;
}

.soh-drill-down-station-label {
  font-size: 1.5em;
  margin-top: 0.25em;
  padding-left: 0.125em;
  text-rendering: geometricPrecision;
}

.display-title {
  padding: 0.25em 0 0 0.25em;
  display: flex;
  align-items: baseline;
  &__subtitle {
    margin-left: 0.5em;
    color: var(--gms-prominent, $gms-prominent);
    font-size: 1.1rem;
    font-weight: 100;
    @include ellipsis;
  }
}

.soh-table-label {
  padding: 0.5em 0 0.25em 0;
  color: var(--gms-prominent, $gms-prominent);
  margin-left: 0.25em;
  font-size: 1.1rem;
  font-weight: 400;
  font-family: $gms-sans;
  -webkit-font-smoothing: antialiased;
}

.drop-zone__wrapper .drop-zone {
  &--highlighted {
    background-color: var(--gms-table-selection, $gms-table-selection);
  }
}
