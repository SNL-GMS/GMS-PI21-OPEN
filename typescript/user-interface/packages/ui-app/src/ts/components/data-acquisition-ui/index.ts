import * as DataAcquisitionComponents from './components';

export { DataAcquisitionComponents };
