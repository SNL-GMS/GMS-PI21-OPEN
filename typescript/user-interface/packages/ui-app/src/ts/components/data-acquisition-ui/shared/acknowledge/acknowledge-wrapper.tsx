import type { SohTypes } from '@gms/common-model';
import {
  useAcknowledgeSohStatus,
  useAppSelector,
  useGetSohConfigurationQuery
} from '@gms/ui-state';
import { UILogger } from '@gms/ui-util';
import * as React from 'react';
import { toast } from 'react-toastify';

import { isAcknowledgeEnabled } from '../table/utils';
import type { AcknowledgeWrapperProps } from './types';

const logger = UILogger.create('GMS_LOG_SOH_ACKNOWLEDGE', process.env.GMS_LOG_SOH_ACKNOWLEDGE);

/**
 * Creates a clone of the wrapped component, and injects the acknowledgeSohStatus
 * function into the child's props.
 */
export function AcknowledgeWrapper(props: React.PropsWithChildren<AcknowledgeWrapperProps>) {
  const { sohStatus, children } = props;

  // Using React Ref to make sure the sohStationStaleMs value is not captured
  // as undefined when the FunctionComponent is created
  const query = useGetSohConfigurationQuery();
  const sohStationStaleMsRef = React.useRef(undefined);
  React.useEffect(() => {
    sohStationStaleMsRef.current = query.data?.sohStationStaleMs;
  }, [query.data?.sohStationStaleMs]);

  const callAcknowledgeMutation = useAcknowledgeSohStatus();
  const userSessionState = useAppSelector(state => state.app.userSession);
  /**
   * Call the mutation function and save the new state to the backend.
   *
   * @param stationIds modified station ids
   * @param comment (optional) an optional comment for the acknowledgement
   */
  const acknowledgeStationsByName = (stationNames: string[], comment?: string) => {
    // If station entries are not stale or already acknowledged then
    // call acknowledge mutation with station names
    if (
      isAcknowledgeEnabled(
        stationNames,
        sohStatus.stationAndStationGroupSoh.stationSoh,
        sohStationStaleMsRef.current
      )
    ) {
      try {
        const ackArgs: SohTypes.AcknowledgeSohStatus = {
          stationNames,
          comment,
          userName: userSessionState.authenticationStatus.userName
        };
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        callAcknowledgeMutation(ackArgs);
      } catch (e) {
        logger.warn(`Failed to send SOH Acknowledgement.`);
      }
    } else {
      toast.warn('Cannot acknowledge due to stale SOH data');
    }
  };

  /**
   * We store the child in a variable so we can check its type with the
   * isValidElement type guard below.
   */
  const child: React.ReactNode = children;

  /**
   * Verify that the children are indeed a React Element (not just a node)
   */
  if (React.isValidElement(child)) {
    /**
     * React.cloneElement is used to inject the new props into the child. This injects the
     * acknowledgeStationsByName function as a prop into the child.
     * CloneElement should be reasonably performant.
     * See https://stackoverflow.com/questions/54922160/react-cloneelement-in-list-performance
     */
    return <>{React.cloneElement(child, { acknowledgeStationsByName })}</>;
  }
  return undefined;
}
