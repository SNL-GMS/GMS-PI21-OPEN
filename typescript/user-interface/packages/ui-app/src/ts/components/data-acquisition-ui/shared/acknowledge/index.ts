export { AcknowledgeOverlay } from './acknowledge-overlay';
export { WithAcknowledgeProps } from './types';
export { WithAcknowledge } from './with-acknowledge';
