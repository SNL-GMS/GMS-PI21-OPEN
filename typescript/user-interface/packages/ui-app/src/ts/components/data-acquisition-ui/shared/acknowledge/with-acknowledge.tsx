import * as React from 'react';

import { ReduxAcknowledgeContainer } from './acknowledge-container';

/**
 * Adds the acknowledgeSohStatus function to the component provided
 *
 * @param WrappedComponent the component to which we should add the acknowledgeSohStatus function
 */
export function WithAcknowledge<T>(
  WrappedComponent: React.FunctionComponent | React.ComponentClass
): React.FunctionComponent<T> {
  function WithAcknowledgeComponent(props) {
    return (
      <ReduxAcknowledgeContainer>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <WrappedComponent {...props} />
      </ReduxAcknowledgeContainer>
    );
  }
  return WithAcknowledgeComponent;
}
