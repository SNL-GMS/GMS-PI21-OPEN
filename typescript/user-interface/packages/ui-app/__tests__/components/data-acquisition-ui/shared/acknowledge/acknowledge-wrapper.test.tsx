import { MILLISECONDS_IN_SECOND, SECONDS_IN_MINUTES } from '@gms/common-util';
import { getStore, withReduxProvider } from '@gms/ui-state';
import { sohConfiguration } from '@gms/ui-state/__tests__/__data__/soh-configuration-query-data';
import { waitForComponentToPaint } from '@gms/ui-state/__tests__/test-util';
import { render } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import { AcknowledgeWrapper } from '../../../../../src/ts/components/data-acquisition-ui/shared/acknowledge/acknowledge-wrapper';
import type {
  AcknowledgeWrapperProps,
  WithAcknowledgeProps
} from '../../../../../src/ts/components/data-acquisition-ui/shared/acknowledge/types';
import { WithAcknowledge } from '../../../../../src/ts/components/data-acquisition-ui/shared/acknowledge/with-acknowledge';
import { stationAndStationGroupSohStatus } from '../../../../__data__/data-acquisition-ui/soh-overview-data';
import { useQueryStateResult } from '../../../../__data__/test-util-data';

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
const Enzyme = require('enzyme');

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

// eslint-disable-next-line import/no-deprecated
const lodash = jest.requireActual('lodash');
lodash.uniqueId = () => '1';

Date.now = jest.fn().mockReturnValue(() => MILLISECONDS_IN_SECOND * SECONDS_IN_MINUTES);

function flushPromises(): any {
  return new Promise(resolve => {
    setTimeout(resolve, 0);
  });
}

// eslint-disable-next-line @typescript-eslint/no-magic-numbers
jest.setTimeout(60000);

describe('Acknowledge Wrapper', () => {
  it('should be defined', () => {
    expect(AcknowledgeWrapper).toBeDefined();
  });

  class TestClass extends React.PureComponent<Partial<WithAcknowledgeProps>> {
    public render() {
      return <div>Test acknowledge wrapper</div>;
    }
  }

  const TestWithAck = WithAcknowledge(TestClass);
  const store = getStore();

  const Wrapper = withReduxProvider(TestWithAck);

  it('matches a snapshot', () => {
    const { container } = render(<Wrapper />);
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  const sohConfigurationQuery = useQueryStateResult;
  sohConfigurationQuery.data = sohConfiguration;
  const mockAcknowledge = jest.fn().mockReturnValue(new Promise(jest.fn()));
  const mockAckProps: AcknowledgeWrapperProps = {
    sohStatus: {
      loading: false,
      stationAndStationGroupSoh: stationAndStationGroupSohStatus,
      lastUpdated: 10,
      isStale: false
    },
    sohConfigurationQuery
  };

  const ackWrapper = Enzyme.mount(
    <Provider store={store}>
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <AcknowledgeWrapper {...mockAckProps}>
        <TestClass />
      </AcknowledgeWrapper>
    </Provider>
  );

  it.skip('should call mutation when acknowledgeStationsByName is called', async () => {
    await waitForComponentToPaint(ackWrapper);
    ackWrapper.update();
    ackWrapper.find('TestClass').props().acknowledgeStationsByName(['H05N', 'H06N']);
    expect(mockAcknowledge).toHaveBeenCalledTimes(1);
    expect(mockAcknowledge).toHaveBeenCalledWith({
      variables: { stationNames: ['H05N', 'H06N'] }
    });
    flushPromises();
  });

  it.skip('should log an error when mutation fails', () => {
    const spyError = jest.spyOn(console, 'error').mockImplementation();
    const errorMessage = 'got a failed promise';
    const rejection = Promise.reject(errorMessage);
    const mockReject = jest.fn().mockReturnValueOnce(rejection);
    ackWrapper.setProps({
      acknowledgeSohStatus: mockReject
    });
    ackWrapper.update();
    ackWrapper.find('TestClass').props().acknowledgeStationsByName(['H05N', 'H06N']);
    // await waitForComponentToPaint(ackWrapper);
    expect(spyError).toMatchSnapshot();
    spyError.mockRestore();
  });
});
