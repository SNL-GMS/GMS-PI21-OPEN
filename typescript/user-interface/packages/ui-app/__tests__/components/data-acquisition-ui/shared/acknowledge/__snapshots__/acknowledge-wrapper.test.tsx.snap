// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Acknowledge Wrapper matches a snapshot 1`] = `
<div>
  <div>
    Test acknowledge wrapper
  </div>
</div>
`;

exports[`Acknowledge Wrapper should log an error when mutation fails 1`] = `[MockFunction]`;
